tasklist/qa/integration-tests/src/test/java/io/camunda/tasklist/graphql/ProcessSaasIT.java

```diff
@@ -51,35 +53,74 @@ public void shouldReturnOnlyMostRecentVersion() throws IOException {
   }
 
   @Test
+  @SuppressWarnings("unchecked")
   public void shouldReturnAllProcessesBasedOnQueries() throws IOException {
     tester.deployProcess("simple_process.bpmn").waitUntil().processIsDeployed();
     tester.deployProcess("simple_process_2.bpmn").waitUntil().processIsDeployed();
     tester.deployProcess("userTaskForm.bpmn").waitUntil().processIsDeployed();
+    final var response = tester.getAllProcesses("");
+    final List<Map<String, Object>> processes =
+        (List<Map<String, Object>>) response.getRaw("$.data.processes");
+    final List<String> processIds = processes.stream().map(p -> (String) p.get("id")).toList();
 
+    assertEquals(3, processIds.size());
     LOGGER.info("Should return all processes based on empty search query");
     testProcessRetrieval("", 3);
 
-    LOGGER.info("Should return all process based on Process named query");
+    LOGGER.info("Should return process based on Process named query");
     final GraphQLResponse responseProcessNamed = testProcessRetrieval("iMpLe", 1);
     assertEquals("Simple process", responseProcessNamed.get("$.data.processes[0].name"));
 
-    LOGGER.info("Should return all process based on Process definition id query");
+    LOGGER.info("Should return process based on Process definition id query");
     final GraphQLResponse responseProcessDefinition = testProcessRetrieval("FoRm", 1);
     assertEquals(
         "userTaskFormProcess",
         responseProcessDefinition.get("$.data.processes[0].processDefinitionId"));
 
-    LOGGER.info("Should return all process based on Process id query");
-    testProcessRetrieval("2251799813685250", 1);
+    LOGGER.info("Should return process based on Process id query");
+    for (final String processId : processIds) {
+      final GraphQLResponse processResponse = testProcessRetrieval(processId, 1);
+      assertEquals(processId, processResponse.get("$.data.processes[0].id"));
+    }
 
-    LOGGER.info("Should return all process based on partial Process id query");
-    testProcessRetrieval("799813685", 0);
+    LOGGER.info("Should not return all process based on partial Process id query");
+    final String commonProcessSubstring =
+        longestCommonSubstring(processIds.get(0), processIds.get(1), processIds.get(2));
+    testProcessRetrieval(commonProcessSubstring, 0);
 
     LOGGER.info("Should not return");
     testProcessRetrieval("shouldNotReturn", 0);
   }
 
-  private GraphQLResponse testProcessRetrieval(String query, int expectedCount) throws IOException {
+  private String longestCommonSubstring(
+      final String firstProcessId, final String secondProcessId, final String thirdProcessId) {
+    int maxLength = 0;
+    int endIndex = 0;
+
+    final int[][][] charIndex =
+        new int[firstProcessId.length() + 1][secondProcessId.length() + 1]
+            [thirdProcessId.length() + 1];
+
+    for (int i = 1; i <= firstProcessId.length(); i++) {
```

Comment:
💭  Is there a way to reduce the complexity over those loops? Or even divide the code in order to have a more comprehensive understanding? 

I understand we are going over the processes ids in order to have the longest one, but took me sometime to go over the logic.

Maybe extract this to a method.